src/hotspot/os/linux/os_linux.cpp

```diff
@@ -4856,11 +4856,19 @@ uint os::processor_id() {
 }
 
 void os::set_native_thread_name(const char *name) {
+  char buf[16]; // according to glibc manpage, 16 chars incl. '/0'
+  // We may need to truncate the thread name. Since a common pattern is
+  // for thread names is to be both longer than 15 chars and have a trailing number
+  // ("DispatcherWorkerThread21", "C2 CompilerThread#54" etc), we truncate "smartly"
+  // by attempting to preserve the trailing number in the name if there is one
+  // (e.g. "DispatcherWorkerThread21" -> "DispatcherW..21").
+  StringUtils::truncate_middle(name, buf, sizeof(buf));
+  // set name in kernel
+  int rc = prctl(PR_SET_NAME, buf);
+  assert(rc == 0, "prctl(PR_SET_NAME) failed");
   if (Linux::_pthread_setname_np) {
-    char buf [16]; // according to glibc manpage, 16 chars incl. '/0'
-    (void) os::snprintf(buf, sizeof(buf), "%s", name);
-    buf[sizeof(buf) - 1] = '\0';
-    const int rc = Linux::_pthread_setname_np(pthread_self(), buf);
+    // set name in pthread lib
+    rc = Linux::_pthread_setname_np(pthread_self(), buf);
```

Comment:
This seems redundant given we have to do the `prctl` directly ourselves.

Reply:
`pthread_setname_np` does more than `prctl`, though. It also populates the task name entry in /proc. See `ls /proc/<pid>/task/comm | xargs cat` . 

We could do this ourselves, too, admittedly. That could get rid of the dlsym for the _pthread_setname, but goes a bit beyond the scope of this PR. And if glibc ever adds anything to `pthread_setname_np`, we will miss out. What do you think?